AirCasting/MicrophoneSession/MicrophoneManager.swift

```diff
@@ -21,86 +18,110 @@ class MicrophoneManager: ObservableObject {
-                }
+        if audioSession.recordPermission == .granted {
+            addInterruptionsObserver()
+            
+            self.session = session
+            createDBStream()
+            locationProvider.requestLocation()
+            
+            let url = URL(fileURLWithPath: "/dev/null", isDirectory: true)
+            
+            let recordSettings: [String: Any] = [
+                AVFormatIDKey: NSNumber(value: kAudioFormatAppleLossless),
+                AVSampleRateKey: 44100.0,
+                AVNumberOfChannelsKey: 1,
+                AVEncoderAudioQualityKey: AVAudioQuality.min.rawValue
+            ]
+            
+            do {
+                try audioSession.setCategory(AVAudioSession.Category.playAndRecord)
+                try audioSession.setActive(true)
+                try recorder = AVAudioRecorder(url:url, settings: recordSettings)
+                
+            } catch {
+                return
             }
-        }
-        
-        self.session = session
-        createDBStream()
-        locationProvider.requestLocation()
-        
-        let url = URL(fileURLWithPath: "/dev/null", isDirectory: true)
-        
-        let recordSettings: [String: Any] = [
-            AVFormatIDKey: NSNumber(value: kAudioFormatAppleLossless),
-            AVSampleRateKey: 44100.0,
-            AVNumberOfChannelsKey: 1,
-            AVEncoderAudioQualityKey: AVAudioQuality.min.rawValue
-        ]
-        
-        do {
-            try audioSession.setCategory(AVAudioSession.Category.playAndRecord)
-            try audioSession.setActive(true)
-            try recorder = AVAudioRecorder(url:url, settings: recordSettings)
             
-        } catch {
-            return
+            recorder.isMeteringEnabled = true
+            recorder.record()
+            isRecording = true
+            
+            levelTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(getMeasurement), userInfo: nil, repeats: true)
         }
-        
-        recorder.isMeteringEnabled = true
-        recorder.record()
-        isRecording = true
-        
-        levelTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(getMeasurement), userInfo: nil, repeats: true)
     }
     
     @objc func getMeasurement() {
+        if !recorder.isRecording {
+            recorder.record()
+        }
         recorder.updateMeters()
-        
         let level = recorder.averagePower(forChannel: 0)
+        createMeasurement(value: level)
+    }
+    
+    func stopRecording() {
+        levelTimer.invalidate()
+        isRecording = false
+        recorder.stop()
+        recorder = nil
+        session?.status = Int16(SessionStatus.FINISHED.rawValue)
+        do {
+            try context.save()
+        } catch {
+            print("Error for saving session when stopping mic recording: \(error)")
```

Comment:
It would be better to propagate this error further. Its not a good idea to have silent failure. Print may be a solution for some cases where the failure can be ignored. in this case if the context will not be saved then the whole process would make no sense,  the whole manangeobjectcontex also can be polluted with object that also in the future can fail because ignored save here will cause all the next save to fail. It should be clear form the top that a process can fail. Please mark the method `throws`.  If `stopRecording` is called directly in the ui then calll `try! stopRecording()` with a force unwrap try. [Fail-fast](https://dzone.com/articles/fail-fast-principle-in-software-development) also applies here but the call methods like this that could fail should propagate the error because this buiness logic may be used in other places where in theory it would be ok to let it silently fail.